### PR TITLE
Add EmptyContent Component and revenue error/empty states.

### LINKED ISSUE #302: Component: EmptyState/EmptyContent
Design: [inVision](https://automattic.invisionapp.com/d/main#/projects/prototypes/14707693).
Links: See p6riRB-3ja-p2. Calypso has some [prior art](https://wpcalypso.wordpress.com/devdocs/design/empty-content).

screen shot 2018-08-16 at 9 33 59 am

A component is needed to display in instances where no data exists (no orders, etc).

### PR DESCRIPTION
**Pull request by emdashcodes**
Closes #302. Also a part of #316 and #225.

This PR adds a new `EmptyContent` component, and an `ImageAsset` component.

`ImageAsset` provides an easy way to load plugin/WC images relative to the plugin folder.

`EmptyContent` is a component, detailed at p6riRB-3ja-p2, that displays a message with an optional illustration and optional action buttons.

This PR also implements the `EmptyContent` component on the revenue report for both empty/no result states, and if an error occurs.

screen shot 2018-09-03 at 2 51 44 pm

screen shot 2018-09-03 at 2 52 01 pm

I imagine we want to tweak the messages a bit, and use a different illustration for the error. cc @jameskoster.

To Test:
* Visit the revenue report. Run a query that returns no results.
* Switch to a different branch of `woocommerce` to make the API unavailable, and then test the error handling.

### CODE REVIEW

File: client/components/empty-content/index.js [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+/**
+ * A component to be used when there is no data to show.
+ * It can be used as an opportunity to provide explanation or guidance to help a user progress.
+ */
+class EmptyContent extends Component {
+	renderIllustration() {
+		const { illustrationWidth, illustrationHeight } = this.props;
+		let illustrationSrc = this.props.illustration;
+		let ImageComponent = 'img';
+		if ( startsWith( illustrationSrc, '/' ) ) {
```
Comment (psealock) [inline client/components/empty-content/index.js]: Can we move this logic to `<ImageAsset />` ? If we alway use it, instead of `` we won't have to replicate this bit of code

File: client/components/empty-content/index.js [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+				<Button className="woocommerce-empty-content__action" isPrimary onClick={ actionCallback }>
+					{ actionLabel }
+				</Button>
+			);
+		}
+
+		return null;
+	}
+
+	renderSecondaryAction() {
```
Comment (psealock) [inline client/components/empty-content/index.js]: Combine `renderSecondaryAction` and `renderPrimaryAction` into one function?

File: client/components/empty-content/index.js [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+		}
+
+		return null;
+	}
+
+	renderActions() {
+		const { actionLabel, secondaryActionLabel } = this.props;
+
+		if ( ! actionLabel && ! secondaryActionLabel ) {
+			return null;
```
Comment (psealock) [inline client/components/empty-content/index.js]: If we make these required proptypes, it'll be a good reminder for devs

File: client/components/image-asset/index.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+ImageAsset.propTypes = {
+	/**
+	 * Image location to pass through to `<img />`.
+	 */
+	src: PropTypes.string,
+	/**
+	 * Alt text to pass through to `<img />`.
+	 */
+	alt: PropTypes.string,
```
Comment (psealock) [inline client/components/image-asset/index.js]: isRequired?

**Review by psealock — COMMENTED** [review]
Nice work! 

When an api error occurs, I get an error from the uitl function `getAllReportData`.

screen shot 2018-09-04 at 11 30 06 am

File: client/components/empty-content/index.js (hunk 3 above)
Comment (justinshreve) [reply client/components/empty-content/index.js]: I made the primary action required. I think we can leave the secondary one as optional.

**Comment by justinshreve** [comment]
Thanks for the review @psealock! I handled all your feedback, and added some additional handling to `getAllReportData` that should help with that error. If you are still getting a JS error, can you paste what the API request and response is?

File: lib/client-assets.php [hunk 5]
```diff
@@ -57,10 +57,16 @@ function wc_admin_register_script() {
 
 	wp_enqueue_script( 'wp-api' );
 
+	/**
+	 * TODO: On merge, once plugin images are added to core WooCommerce, `wcAdminAssetUrl` can be retired, and 
+	 * wcAdminAssetUrl can be used in its place throughout the codebase.
```
Comment (ryelle) [inline lib/client-assets.php]: Should this say that `wcAdminAssetUrl` will be replaced with `wcAssetUrl`?

File: client/components/image-asset/index.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+/**
+ * A component that loads an image, allowing images to be loaded relative to the main asset/plugin folder.
+ * Props are passed through to `<img />`
+ */
+class ImageAsset extends Component {
+	render() {
+		const { src, alt, ...restOfProps } = this.props;
+		let illustrationSrc = src;
+
+		if ( startsWith( illustrationSrc, '/' ) ) {
```
Comment (ryelle) [inline client/components/image-asset/index.js]: Why not use `illustrationSrc.indexOf( '/' ) === 0`?

File: client/components/empty-content/index.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+			'secondary' === type ? this.props.secondaryActionLabel : this.props.actionLabel;
+		const actionURL = 'secondary' === type ? this.props.secondaryActionURL : this.props.actionURL;
+		const actionCallback =
+			'secondary' === type ? this.props.secondaryActionCallback : this.props.actionCallback;
+
+		if ( actionURL && actionCallback ) {
+			return (
+				<Button
+					className="woocommerce-empty-content__action"
+					isPrimary
```
Comment (ryelle) [inline client/components/empty-content/index.js]: Should this depend on whether it's `primary` or `secondary`?

File: client/components/empty-content/index.js [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+	 */
+	title: PropTypes.string.isRequired,
+	/**
+	 * An additional message to be displayed.
+	 */
+	message: PropTypes.string,
+	/**
+	 * The url string of an image path. Prefix with `/` to load an image relative to the plugin directory.
+	 */
+	illustration: PropTypes.string.isRequired,
```
Comment (ryelle) [inline client/components/empty-content/index.js]: What do you think about adding `empty-content.svg` as a default image here, and have this not be required? Do you think there's a case for having EmptyContent without an image?

File: client/components/empty-content/style.scss [hunk 9]
```diff
@@ -0,0 +1,10 @@
+/** @format */
+
+.woocommerce-empty-content {
+	text-align: center;
+
+	.woocommerce-empty-content__action.is-primary {
+		color: $white;
```
Comment (ryelle) [inline client/components/empty-content/style.scss]: Why is this necessary, shouldn't the button already be white?

File: client/components/empty-content/style.scss [hunk 10]
```diff
@@ -0,0 +1,10 @@
...
+		margin-right: $gap;
```
Comment (ryelle) [inline client/components/empty-content/style.scss]: This is adding an offset when there's just a single button, you could do this instead to make sure you're putting space between two buttons:

```
.woocommerce-empty-content__actions {
   .components-button + .components-button {
      margin-left: $gap;
   }
}
```

File: client/components/empty-content/style.scss (hunk 9 above)
Comment (ryelle) [reply client/components/empty-content/style.scss]: nevermind, I see it's a problem when the button is a link - we should fix that in the global scss file, not per-component.

File: images/empty-content.svg [hunk 11]
```diff
@@ -0,0 +1 @@
....2-.4-.3-.7-.4-1.1-.1-.4-.1-.7 0-1.1.1-.7.5-1.4 1-1.8l-.3-.4-.3-.4c-.4.3-.7.7-.9 1.1-.2.4-.3.9-.4 1.4 0 .5.1.9.2 1.4l.3.6c.1.2.3.4.4.5h.3zM16.2 61.3c.5-.5 1.3-.7 2-.7.4 0 .7.1 1.1.2.4.1.7.3 1 .6.1 0 .1.1.2 0 .1 0 .1-.1.1-.2-.2-.4-.6-.7-.9-1-.4-.3-.8-.4-1.3-.5-.5-.1-1 0-1.4.1-.5.1-.9.4-1.3.7l.3.5.2.3zM26 61.4c.3-.3.6-.5 1-.6.3-.1.7-.2 1.1-.2.7 0 1.4.2 2 .7l.3-.4.3-.4c-.4-.3-.8-.5-1.3-.7-.5-.1-1-.2-1.4-.1-.5.1-.9.2-1.3.5s-.7.6-.9 1v.2c.1.1.2.1.2 0z"/><path d="M31.6 61.6c-.2-.4-.5-.8-.9-1.1l-.4.5-.2.3c.6.4.9 1.1 1 1.8.1.4.1.7 0 1.1-.1.4-.2.7-.4 1.1v.2c.1.1.2.1.2 0 .2-.1.3-.3.4-.5l.3-.6c.2-.4.3-.9.2-1.4.2-.5 0-.9-.2-1.4zM25 67.3c-.4.5-.8 1.1-1.1 1.7 0-1.2-.1-2.4-.3-3.6 0-.1-.1-.1-.1-.1-.1 0-.2 0-.2.1-.3 1.8-.3 3.6-.3 5.5 0 .2.1.4.3.4.2.1.5 0 .7-.2.6-1.2 1.1-2.4 1.5-3.6 0-.1 0-.1-.1-.2-.3-.1-.4-.1-.4 0z"/><path d="M23.4 61.7c-4.4 0-8 3.6-8 8 0 2.1.8 4 2.2 5.5-.1.4-.2.7-.3 1.1-.3 1.2-.5 2.5-.5 2.5 0 .1 0 .1.1.2.1 0 .2 0 .2-.1 0 0 .6-1.1 1-2.3.1-.2.2-.4.3-.7 1.4 1.1 3.1 1.7 5 1.7s3.6-.7 5-1.7c.1.2.2.4.3.7.5 1.2 1 2.3 1 2.3 0 .1.1.1.2.1s.2-.1.1-.2c0 0-.2-1.3-.5-2.5-.1-.4-.2-.8-.3-1.1 1.4-1.4 2.2-3.4 2.2-5.5-.1-4.4-3.6-8-8-8zm0 15c-3.9 0-7-3.1-7-7s3.1-7 7-7 7 3.1 7 7-3.2 7-7 7z"/></g></svg>
```
Comment (ryelle) [inline images/empty-content.svg]: Is this meant to be [the person in a wheelchair illustration](https://wpcalypso.wordpress.com/calypso/images/extensions/woocommerce/woocommerce-relaxed.svg)? I see the wheel was removed but the handle wasn't, so the chair itself looks a little odd @jameskoster @LevinMedia (the non-wheelchair version [has a sharper angle chair back](https://wpcalypso.wordpress.com/calypso/images/illustrations/checkEmailsDesktop.svg), too)

**Review by ryelle — COMMENTED** [review]
Looking good - I have a few questions about things & some CSS suggestions.

**Review by timmyc — APPROVED** [review]
Couple of minor comments, and this is testing well for me ( the error case that is since I still can't get the report endpoints to work ). Pre-approving.

File: client/components/empty-content/index.js (hunk 7 above)
Comment (justinshreve) [reply client/components/empty-content/index.js]: Oops yeah. I combined some logic after Paul's review and missed that. Thanks for catching!

File: client/components/empty-content/index.js (hunk 8 above)
Comment (justinshreve) [reply client/components/empty-content/index.js]: I'm fine with that, and it's what the Calypso version does actually. I think most of the time we'll probably want an image but I could see wanting just the text display in some circumstances.

File: client/analytics/report/revenue/index.js [hunk 12]
```diff
@@ -373,27 +374,38 @@ export class RevenueReport extends Component {
...
+			if ( primaryData.isError || secondaryData.isError ) {
+				title = __( 'There was an error getting your stats. Please try again.', 'wc-admin' );
+				actionLabel = __( 'Reload', 'wc-admin' );
+				actionCallback = () => {
+					window.location.reload();
+				};
+			} else {
+				title = __( 'No results could be found for this date range.', 'wc-admin' );
+				actionLabel = __( 'View Orders', 'wc-admin' );
+				actionURL = getAdminLink( 'edit.php?post_type=shop_order' );
```
[the remark this answers is not in the document]
Comment (justinshreve) [reply client/analytics/report/revenue/index.js]: Yeah I wasn't sure exactly what action we would want here so I was considering this temporary.

File: client/analytics/report/revenue/index.js [hunk 13]
```diff
@@ -373,27 +374,38 @@ export class RevenueReport extends Component {
...
-		if ( isReportDataEmpty( primaryData ) ) {
-			return tempMessage( 'Empty Data' );
-		}
+		if ( isReportDataEmpty( primaryData ) || primaryData.isError || secondaryData.isError ) {
+			let title, actionLabel, actionURL, actionCallback;
+			if ( primaryData.isError || secondaryData.isError ) {
+				title = __( 'There was an error getting your stats. Please try again.', 'wc-admin' );
+				actionLabel = __( 'Reload', 'wc-admin' );
+				actionCallback = () => {
+					window.location.reload();
```
[the remark this answers is not in the document]
Comment (justinshreve) [reply client/analytics/report/revenue/index.js]: I'll add a todo for now, since I don't think we currently have any way to track things yet.

**Comment by justinshreve** [comment]
I've handled all the feedback here, except for replacing the illustration/text/link targets, which can be handled in a follow-up once design chimes in.

**Comment by jameskoster** [comment]
> I imagine we want to tweak the messages a bit, and use a different illustration for the error.

On my to-do list :)
